src/cir.rs

```diff
@@ -0,0 +1,451 @@
+// cir:  compact intermediate representation
+use crate::ast::Predicate;
+use crate::context::{Context, Match};
+use crate::interpreter::Execute;
+use crate::lir::{is_operator, LirInstruction, LirLogicalOperator, LirProgram, Translate};
+use crate::semantics::FieldCounter;
+use std::collections::HashMap;
+
+#[derive(Debug)]
+pub struct CirProgram {
+    pub(crate) instructions: Vec<CirInstruction>,
+}
+
+impl CirProgram {
+    pub fn new() -> Self {
+        Self {
+            instructions: Vec::new(),
+        }
+    }
+}
+
+impl Default for CirProgram {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+#[derive(Debug)]
+pub enum CirInstruction {
+    AndIns(AndIns),
+    OrIns(OrIns),
+    NotIns(NotIns),
+    Predicate(Predicate),
+}
+
+#[derive(Debug)]
+pub struct AndIns {
+    left: CirOperand,
+    right: CirOperand,
+}
+
+#[derive(Debug)]
+pub struct OrIns {
+    left: CirOperand,
+    right: CirOperand,
+}
+
+#[derive(Debug)]
+pub struct NotIns {
+    right: CirOperand,
+}
+
+#[derive(Debug, Clone)]
+pub enum CirOperand {
+    Index(usize),
+    Predicate(Predicate),
+}
+
+impl CirOperand {
+    #[inline]
+    pub fn as_predicate(&self) -> &Predicate {
+        match &self {
+            CirOperand::Index(_index) => {
+                panic!("Call as_predicate on index Operand, CirProgram is wrong.")
+            }
+            CirOperand::Predicate(p) => p,
+        }
+    }
+
+    #[inline]
+    pub fn as_index(&self) -> usize {
+        match &self {
+            CirOperand::Index(index) => *index,
+            CirOperand::Predicate(_p) => {
+                panic!("Call as_index on predicate Operand, CirProgram is wrong.")
+            }
+        }
+    }
+}
+
+#[inline]
+fn is_index(cir_operand: &CirOperand) -> bool {
+    match cir_operand {
+        CirOperand::Index(_index) => true,
+        CirOperand::Predicate(_p) => false,
+    }
+}
```

Comment:
I'm surprised that Rust does not have a mechanism to assume or assert an enum variant.

Reply:
Because that assertion isn't necessary to be a part of the language itself. `match` and `unreachable` are enough to do this right. Perhaps some third-party crates may provide macros like `#[derive(unwrap_into_variant)]` to simplify this. But I don't think it's very needed for an external crate to do this simple job.